### PR TITLE
refactor(gatsby-transformer-sharp): Use explicit graphql type definition vs inferring

### LINKED ISSUE #17531: [gatsby schema] Convert gatsby-transformer-sharp to use schema customization
Currently `gatsby-transformer-sharp` uses inference and old Gatsby schema customization (`extend-node-type`) for adding fields. There are several obscure errors that happen because of this.

It would be great to move sharp to use new schema customization. We should then also bump the version of transformer and make it rely on latest gatsby that has all the schema customization features.

If there are some features in schema customization that are missing for it to be possible, we should add that to schema customization along the way.

### PR DESCRIPTION
**Pull request by vladar**
## Description

Convert gatsby-transformer-sharp to use schema customization API (vs. relying on type inference)

Fixes #17531

### CODE REVIEW

File: packages/gatsby-transformer-sharp/src/customize-schema.js [hunk 1]
```diff
@@ -546,3 +538,35 @@ module.exports = ({
...
+    name: `ImageSharp`,
+    fields: createFields({
+      pathPrefix,
+      getNodeAndSavePathDependency,
+      reporter,
+      cache,
+    }),
+    interfaces: [`Node`],
+    extensions: {
+      infer: true,
```
Comment (freiksenet) [inline packages/gatsby-transformer-sharp/src/customize-schema.js]: Let's actually try to get it to `infer: false`. In addition to fields that we add in `createFields`, could you look at sourceNodes code for this plugin and see what fields are created to be inferred?

Comment (vladar) [reply packages/gatsby-transformer-sharp/src/customize-schema.js]: There was a discussion that we should keep it for BC but also add the `deprecated` version of `infer` which would warn when new node fields were added without corresponding schema extension.

Comment (freiksenet) [reply packages/gatsby-transformer-sharp/src/customize-schema.js]: Ok, we don't have to block this based on this. Could you add all the existing fields we add in the plugin and then we can merge it with infer: true, with a note that we'll deprecate it when deprecated infer lands.

Comment (vladar) [reply packages/gatsby-transformer-sharp/src/customize-schema.js]: @freiksenet I compared schemas generated with `infer: false` and `infer: true` and they are already identical, so I assume this plugin already describes all the fields

Comment (freiksenet) [reply packages/gatsby-transformer-sharp/src/customize-schema.js]: OK, then I was confused about this :) Sorry about it. This is good to be merged.

File: packages/gatsby-transformer-sharp/src/gatsby-node.js [hunk 2]
```diff
@@ -1,32 +1,14 @@
...
 
-  // Check if there are any ImageSharp nodes. If so add fragments for ImageSharp.
-  // The fragment will cause an error if there are no ImageSharp nodes.
-  if (getNodesByType(`ImageSharp`).length == 0) {
-    return
-  }
-
-  // We have ImageSharp nodes so let's add our fragments to .cache/fragments.
+  // Add fragments for ImageSharp to .cache/fragments.
   await fs.copy(
```
Comment (sidharthachatterjee) [inline packages/gatsby-transformer-sharp/src/gatsby-node.js]: This is no longer necessary as we run babel on plugin code now. We should be able to remove the copy operation.

File: packages/gatsby-transformer-sharp/src/gatsby-node.js [hunk 3]
```diff
@@ -1,32 +1,14 @@
 const fs = require(`fs-extra`)
 
 exports.onCreateNode = require(`./on-node-create`)
-exports.setFieldsOnGraphQLNodeType = require(`./extend-node-type`)
+exports.createSchemaCustomization = require(`./customize-schema`)
```
Comment (sidharthachatterjee) [inline packages/gatsby-transformer-sharp/src/gatsby-node.js]: Need to bump dependency on `gatsby` since older versions won't have this API.

Comment (vladar) [reply packages/gatsby-transformer-sharp/src/gatsby-node.js]: Added in https://github.com/gatsbyjs/gatsby/pull/18871/commits/2fb1c2061f52ef0b934c8fd6c5e3b7fb52c05406

**Comment by sidharthachatterjee** [comment]
Published in `gatsby-transformer-sharp@2.3.0`

**Pull request merged**